src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+/-- Retrieves the mapping for the given key. Ensures that such a mapping exists by requiring a proof
+of `a ∈ m`.
+
+Uses the `LawfulBEq` instance to cast the retrieved value to the correct type. -/
+@[inline] def get [BEq α] [Hashable α] [LawfulBEq α] (m : Raw α β) (a : α) (h : a ∈ m) : β a :=
+  Raw₀.get ⟨m, by change dite .. = true at h; split at h <;> simp_all⟩ a
+    (by change dite .. = true at h; split at h <;> simp_all)
+
+/-- Tries to retrieve the mapping for the given key, returning `fallback` if no such mapping is
+present.
+
+Uses the `LawfulBEq` instance to cast the retrieved value to the correct type. -/
+@[inline] def getD [BEq α] [Hashable α] [LawfulBEq α] (m : Raw α β) (a : α) (fallback : β a) :
+    β a :=
+  if h : 0 < m.buckets.size then
+    Raw₀.getD ⟨m, h⟩ a fallback
+  else fallback -- will never happen for well-formed inputs
+
+/-- Tries to retrieve the mapping for the given key, panicking if no such mapping is present.
+
+Uses the `LawfulBEq` instance to cast the retrieved value to the correct type. -/
+@[inline] def get! [BEq α] [Hashable α] [LawfulBEq α] (m : Raw α β) (a : α) [Inhabited (β a)] :
+    β a :=
+  if h : 0 < m.buckets.size then
+    Raw₀.get! ⟨m, h⟩ a
+  else default -- will never happen for well-formed inputs
+
+/-- Removes the mapping for the given key if it exists.
+
+Uses the `LawfulBEq` instance to cast the retrieved value to the correct type. -/
+@[inline] def remove [BEq α] [Hashable α] (m : Raw α β) (a : α) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.remove ⟨m, h⟩ a
+  else m -- will never happen for well-formed inputs
+
+section
+
+variable {β : Type v}
+
+/-- Tries to retrieve the mapping for the given key, returning `none` if no such mapping is present.
+-/
+@[inline] def Const.get? [BEq α] [Hashable α] (m : Raw α (fun _ => β)) (a : α) : Option β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.Const.get? ⟨m, h⟩ a
+  else none -- will never happen for well-formed inputs
+
+/-- Retrieves the mapping for the given key. Ensures that such a mapping exists by requiring a proof
+of `a ∈ m`. -/
+@[inline] def Const.get [BEq α] [Hashable α] (m : Raw α (fun _ => β)) (a : α) (h : a ∈ m) : β :=
+  Raw₀.Const.get ⟨m, by change dite .. = true at h; split at h <;> simp_all⟩ a
+    (by change dite .. = true at h; split at h <;> simp_all)
+
+/-- Tries to retrieve the mapping for the given key, returning `fallback` if no such mapping is
+present. -/
+@[inline] def Const.getD [BEq α] [Hashable α] (m : Raw α (fun _ => β)) (a : α) (fallback : β) : β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.Const.getD ⟨m, h⟩ a fallback
+  else fallback -- will never happen for well-formed inputs
+
+/-- Tries to retrieve the mapping for the given key, panicking if no such mapping is present. -/
+@[inline] def Const.get! [BEq α] [Hashable α] [Inhabited β] (m : Raw α (fun _ => β)) (a : α) : β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.Const.get! ⟨m, h⟩ a
+  else default -- will never happen for well-formed inputs
```

Comment:
```suggestion
  else default
```
Sure the empty map is well-formed, and I can call `get!` on it

Reply:
The empty map has a non-empty buckets array consisting of empty buckets :) So on the empty map we are taking the true-branch of the `if`.

Reply:
See also the theorem `WF.size_buckets_pos` in that file.

Reply:
Got it, thanks :)